File: app/schemas.py
"""Pydantic request/response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class HealthModelsStatus(BaseModel):
    """Model availability state for health endpoint."""

    cloth_generator: str
    tryon_model: str


class HealthResponse(BaseModel):
    """Health response schema."""

    status: str
    device: str
    requested_device: str
    cuda_available: bool
    mps_available: bool
    models: HealthModelsStatus


class GenerateClothesRequest(BaseModel):
    """Request payload for clothing generation."""

    prompt: str = Field(..., min_length=3)
    category: str | None = None
    negative_prompt: str | None = None
    count: int = Field(default=1, ge=1, le=8)
    width: int = Field(default=1024, ge=256, le=1536)
    height: int = Field(default=1024, ge=256, le=1536)
    guidance_scale: float = Field(default=0.0, ge=0.0, le=20.0)
    num_inference_steps: int = Field(default=4, ge=1, le=100)
    seed: int | None = None


class GenerateClothItem(BaseModel):
    """Single generated cloth item."""

    id: str
    path: str
    prompt: str
    seed: int | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class GenerateClothesResponse(BaseModel):
    """Response for clothing generation."""

    success: bool
    items: list[GenerateClothItem]


class GenerateClothesJobCreateResponse(BaseModel):
    """Response payload for async clothes generation job creation."""

    success: bool
    job_id: str
    status: str


class GenerateClothesJobStatusResponse(BaseModel):
    """Response payload for async clothes generation job status polling."""

    success: bool
    job_id: str
    status: str
    items: list[GenerateClothItem] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    error: str | None = None


class VirtualTryOnPathRequest(BaseModel):
    """Request payload for try-on using local file paths."""

    person_image_path: str
    garment_image_path: str
    category: str = "upper_body"


class VirtualTryOnResponse(BaseModel):
    """Response payload for try-on endpoints."""

    success: bool
    result_path: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class VirtualTryOnJobCreateResponse(BaseModel):
    """Response payload for async try-on job creation."""

    success: bool
    job_id: str
    status: str


class VirtualTryOnJobStatusResponse(BaseModel):
    """Response payload for async try-on job status polling."""

    success: bool
    job_id: str
    status: str
    result_path: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    error: str | None = None

File: frontend/lib/schemas.ts
import { z } from "zod";
import { garmentCategories } from "@/lib/constants";

export const healthModelsStatusSchema = z.object({
  cloth_generator: z.enum(["available", "missing"]),
  tryon_model: z.enum(["available", "missing"])
});

export const healthResponseSchema = z.object({
  status: z.string(),
  device: z.string(),
  requested_device: z.string(),
  cuda_available: z.boolean(),
  mps_available: z.boolean(),
  models: healthModelsStatusSchema
});

export const generateClothesRequestSchema = z.object({
  prompt: z.string().min(3, "Prompt must be at least 3 characters long"),
  category: z.string().trim().optional().or(z.literal("")).transform((value) => value || undefined),
  negative_prompt: z
    .string()
    .trim()
    .optional()
    .or(z.literal(""))
    .transform((value) => value || undefined),
  count: z.coerce.number().int().min(1).max(8),
  width: z.coerce.number().int().min(256).max(1536),
  height: z.coerce.number().int().min(256).max(1536),
  guidance_scale: z.coerce.number().min(0).max(20),
  num_inference_steps: z.coerce.number().int().min(1).max(100),
  seed: z.preprocess((value) => {
    if (value === "" || value === null || value === undefined) {
      return undefined;
    }
    return value;
  }, z.coerce.number().int().optional())
});

export const generateClothItemSchema = z.object({
  id: z.string(),
  path: z.string(),
  prompt: z.string(),
  seed: z.number().int().nullable().optional(),
  metadata: z.record(z.unknown())
});

export const generateClothesResponseSchema = z.object({
  success: z.boolean(),
  items: z.array(generateClothItemSchema)
});

export const generateClothesJobCreateResponseSchema = z.object({
  success: z.boolean(),
  job_id: z.string(),
  status: z.string()
});

export const generateClothesJobStatusResponseSchema = z.object({
  success: z.boolean(),
  job_id: z.string(),
  status: z.string(),
  items: z.array(generateClothItemSchema),
  metadata: z.record(z.unknown()),
  error: z.string().nullable().optional()
});

export const virtualTryOnPathRequestSchema = z.object({
  person_image_path: z.string().min(1),
  garment_image_path: z.string().min(1),
  category: z.enum(["upper_body", "lower_body", "dress"])
});

export const virtualTryOnResponseSchema = z.object({
  success: z.boolean(),
  result_path: z.string(),
  metadata: z.record(z.unknown())
});

export const virtualTryOnJobCreateResponseSchema = z.object({
  success: z.boolean(),
  job_id: z.string(),
  status: z.string()
});

export const virtualTryOnJobStatusResponseSchema = z.object({
  success: z.boolean(),
  job_id: z.string(),
  status: z.string(),
  result_path: z.string().nullable().optional(),
  metadata: z.record(z.unknown()),
  error: z.string().nullable().optional()
});

export const tryOnCategoryOptions = garmentCategories;

export type HealthResponse = z.infer<typeof healthResponseSchema>;
export type GenerateClothesRequest = z.infer<typeof generateClothesRequestSchema>;
export type GenerateClothItem = z.infer<typeof generateClothItemSchema>;
export type GenerateClothesResponse = z.infer<typeof generateClothesResponseSchema>;
export type GenerateClothesJobCreateResponse = z.infer<typeof generateClothesJobCreateResponseSchema>;
export type GenerateClothesJobStatusResponse = z.infer<typeof generateClothesJobStatusResponseSchema>;
export type VirtualTryOnPathRequest = z.infer<typeof virtualTryOnPathRequestSchema>;
export type VirtualTryOnResponse = z.infer<typeof virtualTryOnResponseSchema>;
export type VirtualTryOnJobCreateResponse = z.infer<typeof virtualTryOnJobCreateResponseSchema>;
export type VirtualTryOnJobStatusResponse = z.infer<typeof virtualTryOnJobStatusResponseSchema>;
